src/oscontainer.py

```diff
@@ -0,0 +1,154 @@
+    # In theory we shouldn't need --entrypoint=/enoent here, but
+    # it works around a podman bug.
+    cid = run_get_string(['podman', rootarg, 'create', '--entrypoint=/enoent', iid])
+    mnt = run_get_string(['podman', rootarg, 'mount', cid])
+    try:
+        src_repo = os.path.join(mnt, 'srv/repo')
+        run_verbose(["ostree", "--repo="+dest, "pull-local", src_repo, commit])
+    finally:
+        subprocess.call(['podman', rootarg, 'umount', cid])
+    if args.ref is not None:
+        run_verbose(["ostree", "--repo="+dest, "refs", '--create='+args.ref, commit])
+
+
+# Given an OSTree repository at src (and exactly one ref) generate an oscontainer
+# with it.
+def oscontainer_build(containers_storage, src, ref, image_name_and_tag,
+                      base_image, push=False, tls_verify=True,
+                      inspect_out=None):
+    r = OSTree.Repo.new(Gio.File.new_for_path(src))
+    r.open(None)
+
+    [_, rev] = r.resolve_rev(ref, True)
+    if ref != rev:
+        print("Resolved {} = {}".format(ref, rev))
+    [_, ostree_commit, _] = r.load_commit(rev)
+    ostree_commitmeta = ostree_commit.get_child_value(0)
+    versionv = ostree_commitmeta.lookup_value("version", GLib.VariantType.new("s"))
+    if versionv:
+        ostree_version = versionv.get_string()
+    else:
+        ostree_version = None
+
+    rootarg = '--root='+containers_storage
+    bid = run_get_string(['buildah', rootarg, 'from', base_image])
+    mnt = run_get_string(['buildah', rootarg, 'mount', bid])
+    try:
+        dest_repo = os.path.join(mnt, 'srv/repo')
+        subprocess.check_call(['mkdir', '-p', dest_repo])
+        subprocess.check_call(["ostree", "--repo="+dest_repo, "init", "--mode=archive"])
+        # Note that oscontainers don't have refs
+        print("Copying ostree commit into container: {} ...".format(rev))
+        run_verbose(["ostree", "--repo="+dest_repo, "pull-local", src, rev])
+
+        # We use /noentry to trick `podman create` into not erroring out
+        # on a container with no cmd/entrypoint.  It won't actually be run.
+        config=['--entrypoint', '["/noentry"]',
+                '-l', OSCONTAINER_COMMIT_LABEL+'='+rev]
+        if ostree_version is not None:
+            config += ['-l', 'version='+ostree_version]
+        run_verbose(['buildah', rootarg, 'config'] + config + [bid])
+        print("Committing container...")
+        iid = run_get_string(['buildah', rootarg, 'commit', bid, image_name_and_tag])
+        print("{} {}".format(image_name_and_tag, iid))
+    finally:
+        subprocess.call(['buildah', rootarg, 'umount', bid], stdout=subprocess.DEVNULL)
+        subprocess.call(['buildah', rootarg, 'rm', bid], stdout=subprocess.DEVNULL)
+
+    if push:
+        print("Pushing container")
+        if not tls_verify:
+            tls_arg = '--tls-verify=false'
+        else:
+            tls_arg = '--tls-verify'
+        run_verbose(['podman', rootarg, 'push', tls_arg, image_name_and_tag])
```

Comment:
I learned that even doing `podman inspect` will initialize the CNI plugin, which is where the requirement on `nsenter`, `iptables`, and `ip6tables` comes from.

https://github.com/containers/libpod/issues/2384

We could switch to using `skopeo`, but we still have a requirement on needing storage in the appliance for the build operation, so it wouldn't really change much overall.

Reply:
Hm, interesting point.  We use `podman` elsewhere but maybe we could drop it?  If `buildah` doesn't bring in iptables then that's good.

Reply:
@cgwalters so basically the only thing we would need podman for is the `podman inspect` - can we get that information another way?

Reply:
FWIW, `skopeo inspect` doesn't like containers storage outside of `/var/lib/containers/storage`

https://github.com/containers/skopeo/issues/547#issuecomment-465397396

And `buildah inspect` has a different format than `podman inspect`

Reply:
I got output from `skopeo inspect` but i'm guessing this isn't sufficient:

```
bash-4.4# skopeo inspect containers-storage:localhost/fcos:29.43
[  331.745337] overlayfs: unrecognized mount option "nodev" or missing value
WARN[0000] Not using native diff for overlay, this may cause degraded performance for building images: failed to mount overlay: invalid argument
{
    "Name": "localhost/fcos",
    "Digest": "sha256:2daf187de979579501c6a97e773c32d8e904428df1a637a9ee50e8741219d120",
    "RepoTags": [],
    "Created": "2019-02-21T21:29:17.05974917Z",
    "DockerVersion": "",
    "Labels": {
        "com.coreos.ostree-commit": "55b7963ceb04b1c6cd9b170cdce39d23dcde108456c81b07564fffb0787f70b0",
        "version": "29.43"
    },
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:d7bae06c300baeb14772f772af4a9f5b21fe1c43eae90de7a0e4bf908eaa8008"
    ]
}
```